app/controllers/application_controller.rb

```diff
@@ -15,17 +16,41 @@ class ApplicationController < ActionController::Base
   end
 
   def current_user
-    user = super
-    if user
-      User.includes(:roles).find(user.id)
-    end
+    super
+
+    # TODO: this breaks lots of assumptions about current_user
+    # being writeable, and requires more thought.
+    # if user
+    #   User.includes(:roles).find(user.id)
+    # end
   end
 
   def decorated_current_user
     current_user.try(:decorate)
   end
 
+  #TODO: move this into a Warden after_authentication method or
+  # something likely to be executed _after_ TFA has been skipped
+  def complete_two_factor_setup
+    if current_user && !current_user.account_verified && current_user.confirmed_at
+      # Users need to be able to verify their tfa code
+      #unless request.path.start_with?('/users')
+      if request.path.match(/\A\/users(?!\/edit).*\z/) == nil
```

Comment:
This regex is confusing. If there's only a few paths that _don't_ require a redirect, it may make more sense to just white list them rather than making assumptions about routing.